cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,420 @@
+					getFunc: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedGetResult, test.expectedGetResponse, test.expectedGetError
+					},
+					listFunc: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+						return test.expectedListResult, test.expectedListResponse, test.expectedListError
+					},
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+
+			fw, err := fwManagerOp.Get(ctx)
+			if want, got := test.expectedError, err; !reflect.DeepEqual(want, got) {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+
+			if !cmp.Equal(fw, test.expectedFirewall) {
+				t.Errorf("got %q, want %q", fw, test.expectedFirewall)
+			}
+		})
+	}
+}
+
+func TestFirewallController_Set(t *testing.T) {
+	testcases := []struct {
+		name         string
+		fwCache      firewallCache
+		inboundRules []godo.InboundRule
+		firewall     *godo.Firewall
+		// Expected DO Firewall API, AddRules request, response values.
+		expectedAddRulesResponse *godo.Response
+		expectedAddRulesError    error
+		// Expected DO Firewall API, Create request, response values.
+		expectedCreateResult   *godo.Firewall
+		expectedCreateResponse *godo.Response
+		expectedCreateError    error
+		// Expected DO Firewall API, Get request, response values.
+		expectedGetResult   *godo.Firewall
+		expectedGetResponse *godo.Response
+		expectedGetError    error
+		expectedError       error
+		// Expected DO Firewall API, List request, response values.
+		expectedListResult   *godo.Firewall
+		expectedListResponse *godo.Response
+		expectedListError    error
+		// Expected DO Firewall API, Update request, response values.
+		expectedUpdateResult   *godo.Firewall
+		expectedUpdateResponse *godo.Response
+		expectedUpdateError    error
+	}{
+		{
+			name:                   "create firewall when cache does not exist (i.e. initial startup)",
+			fwCache:                newFakeFirewallCacheEmpty(),
+			expectedGetResponse:    newFakeNotFoundResponse(),
+			expectedListResponse:   newFakeNotFoundResponse(),
+			expectedListResult:     &godo.Firewall{},
+			expectedCreateResponse: newFakeOKResponse(),
+			expectedCreateResult:   newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+		},
+		{
+			name:                   "failing to update the firewall because it is missing",
+			fwCache:                newFakeFirewallCache(fakeWorkerFWName, diffFakeInboundRule),
+			inboundRules:           []godo.InboundRule{fakeInboundRule},
+			expectedCreateResponse: newFakeNotOKResponse(),
+			expectedCreateError:    errors.New("failed to create firewall after found missing on update"),
+			expectedListResponse:   newFakeOKResponse(),
+			expectedListResult:     newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			expectedUpdateResponse: newFakeNotFoundResponse(),
+			expectedUpdateError:    errors.New("failed to create firewall after found missing on update"),
```

Comment:
Using this very error message here is slightly confusing because it's one that our controller produces, not the godo update API. Let's choose something like `"firewall missing"`.

This also makes it easier to troubleshoot test failures.